### PR TITLE
Fix routing replay split sizes for attention

### PR DESCRIPTION
**Pull request by vivekkalyan**
## Summary
- preserve the full attention token layout when recording routing replay token UID sets
- keep the compacted GDN token UID layout only for GDN replay
- add a unit test that covers the differing attention-vs-GDN replay shapes

## Why
The Megatron routing replay path could fail with `split_with_sizes expects split_sizes to sum exactly ...` when attention replay used compacted GDN token UID sets. Attention needs the original flattened token layout, while GDN uses the compact routed-token layout.

## Validation
- `uv run --with torch --with safetensors --with megatron-core==0.17.0 --with transformers==5.2.0 --group dev pytest tests/unit/test_moe_routing_replay.py tests/unit/test_dedicated_config.py`
- Sky 2x H200 Bonnie Megatron repro against this fix completed 1 training step without the split-size crash
- Stacked LoRA PR smoke run against this branch also completed 1 Megatron training step successfully

### CODE REVIEW

**Comment by FurtherAI** [comment]
Looks good, seems like a misconception from Codex and not caught by tests because they fill the packed seq and don't have padding.
